tests/test_agent_modules/test_tool_factory_openapi.py

```diff
@@ -0,0 +1,369 @@
+            call_args = request_call.call_args
+            assert "https://api.example.com/users/123" in call_args[0]
+
+    @pytest.mark.asyncio
+    async def test_json_body_for_post_methods(self):
+        """Test that JSON body is sent for POST/PUT/PATCH methods."""
+        openapi = {"servers": [{"url": "https://api.example.com"}]}
+        tool_schema = {"properties": {"requestBody": {"type": "object", "properties": {"data": {"type": "string"}}}}}
+
+        # HTTP mocking IS justified
+        with patch("agency_swarm.tools.tool_factory.httpx.AsyncClient") as mock_client:
+            mock_response = MagicMock()
+            mock_response.json.return_value = {"created": True}
+            mock_client.return_value.__aenter__.return_value.request = AsyncMock(return_value=mock_response)
+
+            callback = ToolFactory._create_invoke_for_path("/test", "POST", openapi, tool_schema, "create_test")
+
+            await callback(None, '{"requestBody": {"data": "test"}}')
+
+            # Verify JSON body was sent - this is the real behavior we care about
+            request_call = mock_client.return_value.__aenter__.return_value.request
+            call_kwargs = request_call.call_args[1]
+            assert call_kwargs["json"] == {"data": "test"}
+
+    @pytest.mark.asyncio
+    async def test_response_text_fallback(self):
+        """Test fallback to text response when JSON parsing fails."""
+        openapi = {"servers": [{"url": "https://api.example.com"}]}
+        tool_schema = {"properties": {}}
+
+        # HTTP mocking IS justified
+        with patch("agency_swarm.tools.tool_factory.httpx.AsyncClient") as mock_client:
+            mock_response = MagicMock()
+            mock_response.json.side_effect = Exception("Not JSON")
+            mock_response.text = "Plain text response"
+            mock_client.return_value.__aenter__.return_value.request = AsyncMock(return_value=mock_response)
+
+            callback = ToolFactory._create_invoke_for_path("/test", "GET", openapi, tool_schema, "text_func")
+
+            result = await callback(None, "{}")
+
+            assert result == "Plain text response"
+
+    @pytest.mark.asyncio
+    async def test_validation_error_handling(self):
+        """Test handling of validation errors in parameters."""
+        tool_schema = {
+            "properties": {
+                "parameters": {
+                    "type": "object",
+                    "properties": {"required_field": {"type": "string"}},
+                    "required": ["required_field"],
+                }
+            }
+        }
+
+        callback = ToolFactory._create_invoke_for_path("/test", "POST", {}, tool_schema, "test_func")
+
+        # Test real validation error with missing required field
+        with pytest.raises((ValidationError, Exception)):  # Should raise validation error
+            await callback(None, '{"parameters": {}}')  # Missing required_field
```

Comment:
Update the validation error handling test to supply OpenAPI server metadata and expect the explicit ModelBehaviorError:

```suggestion
        # Provide minimal OpenAPI metadata so the invoke path builds a proper URL
        openapi = {"servers": [{"url": "https://api.example.com"}]}
        callback = ToolFactory._create_invoke_for_path("/test", "POST", openapi, tool_schema, "test_func")

        # Test real validation error with missing required field
        with pytest.raises(ModelBehaviorError, match="Invalid JSON input in parameters"):
            await callback(None, '{"parameters": {}}')  # Missing required_field
```